### PR TITLE
samples: drivers: uart: Add a 'passthrough' example

### PR DESCRIPTION
**Pull request by attie-argentum**
This sample can be used as a "virtual wire", allowing direct access to devices connected to the target processor - for example: GPS, Cellular, etc...

This is also useful as a utility - I've found myself reaching for a rough implementation I've maintained elsewhere many times... no other connectivity options exist for UART, where other interfaces like SPI and I2C have shell commands.

### CODE REVIEW

**Comment by dcpleung** [comment]
Please fix the compliance check errors.

**Comment by attie-argentum** [comment]
@dcpleung - I got most of them already, the rest should be resolved now

**Review by dcpleung — DISMISSED** [review]

**Comment by cfriedt** [comment]
> Really cool! I was hoping to see such a sample so this is great! 
> @cfriedt -- is this sample something that could be done in a maybe more concise/generic way with the devmux driver? If yes, maybe you could help provide some guidance?

It's possible, yes.

The passthrough device and the hardware UART would just need to be in the phandle property of devmux, and then it's just a matter of switching them back and forth.

File: samples/drivers/uart/passthrough/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(passthrough)
+
+set(CMAKE_EXPORT_COMPILE_COMMANDS on)
+zephyr_compile_options(-fdiagnostics-color=always)
```
Comment (thedjnK) [inline samples/drivers/uart/passthrough/CMakeLists.txt:8]: Why?

**Review by thedjnK — DISMISSED** [review]

**Review by thedjnK — COMMENTED** [review]
Also lacking a sample.yml

File: samples/drivers/uart/passthrough/CMakeLists.txt (hunk 1 above)
Comment (attie-argentum) [reply samples/drivers/uart/passthrough/CMakeLists.txt]: No reason, removed. (I copied the `native_tty` sample)

**Comment by attie-argentum** [comment]
> Also lacking a sample.yml

@thedjnK - thanks, fixed!

**Comment by attie-argentum** [comment]
> The passthrough device and the hardware UART would just need to be in the phandle property of devmux, and then it's just a matter of switching them back and forth (a shell command would do)

@cfriedt - I may be misunderstanding, but doesn't the devmux provide a "mux" type "select-one-device" layer? How would it be wired to feed all data between two UARTs?

**Comment by cfriedt** [comment]
> @cfriedt - I may be misunderstanding, but doesn't the devmux provide a "mux" type "select-one-device" layer? How would it be wired to transparently feed all data between two UARTs? (like replacing the uC with a pair of wires)

I think the devmux would really be more appropriate if you wanted to dynamically switch  between the e.g. modem UART and Zephyr's console, so if you wanted to activate access to the modem UART with an escape sequence or something, and then switch back to Zephyr's console later.

File: samples/drivers/uart/passthrough/CMakeLists.txt (hunk 1 above)
Comment (nordicjm) [reply samples/drivers/uart/passthrough/CMakeLists.txt]: Interesting, will post a PR to remove that

**Comment by kartben** [comment]
> I think the devmux would really be more appropriate if you wanted to dynamically switch between the e.g. modem UART and Zephyr's console, so if you wanted to activate access to the modem UART with an escape sequence or something, and then switch back to Zephyr's console later.

Ya I was thinking something like that, with `zephyr,shell-uart` being set to the devmux, and the devmux node containing the console and whatever other UART(s) one may want to spy on / switch to.
Switching back and forth could be done with an escape sequence indeed (but not sure how to capture it on the "non-console" UARTs??). Or maybe with a physical button on the device, a-la KVM?
Might also be nice to have a `devmux` shell command btw, which would allow to "select set" the devmux (with an optional timeout after which it switches back to previous state?) 

Oh well :) would love to see such an example (hint hint -- if any of the above remotely makes sense to you @cfriedt, I might actually give it a try :)), but this is certainly off-topic for this PR (even if that hypothetical devmux-based implementation would demonstrate a similar "passthrough" scenario).

**Comment by attie-argentum** [comment]
Any chance we can progress this PR? (ping @dcpleung, @thedjnK, @nordicjm)

File: samples/drivers/uart/passthrough/sample.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+tests:
+  sample.drivers.uart:
+    integration_platforms:
+      - qemu_x86
+    tags:
+      - serial
+      - uart
+    filter: CONFIG_SERIAL and
+            CONFIG_UART_INTERRUPT_DRIVEN and
+            dt_chosen_enabled("zephyr,shell-uart")
```
Comment (dcpleung) [inline samples/drivers/uart/passthrough/sample.yml:12]: Shouldn't this be `zephyr,console` instead to mirror `DEV_CONSOLE` below? Also another `dt_chosen_enabled` for `uart,passthrough` might be a good idea.

Comment (attie) [reply samples/drivers/uart/passthrough/sample.yml:12]: Good spot, thanks!

Comment (attie-argentum) [reply samples/drivers/uart/passthrough/sample.yml]: Good spot, thanks... fixed!

**Review by dcpleung — DISMISSED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**